hail/python/test/hailtop/batch/test_batch_pool_executor.py

```diff
@@ -0,0 +1,145 @@
+        future = bpe.submit(sleep_forever)
+    was_cancelled = future.cancel()
+    assert was_cancelled
+    assert future.cancelled()
+
+
+def test_result_with_timeout():
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        future = bpe.submit(sleep_forever)
+        try:
+            future.result(timeout=2)
+        except asyncio.TimeoutError:
+            pass
+        else:
+            assert False
+        finally:
+            future.cancel()
+
+
+def test_map_chunksize():
+    row_args = [x
+                for row in range(5)
+                for x in [row, row, row, row, row]]
+    col_args = [x
+                for row in range(5)
+                for x in list(range(5))]
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        multiplication_table = list(bpe.map(lambda x, y: x * y,
+                                            row_args,
+                                            col_args,
+                                            chunksize=5))
+    assert multiplication_table == [
+        0,  0,  0,  0,  0,
+        0,  1,  2,  3,  4,
+        0,  2,  4,  6,  8,
+        0,  3,  6,  9, 12,
+        0,  4,  8, 12, 16]
+
+
+def test_map_timeout():
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        try:
+            list(bpe.map(lambda _: sleep_forever(), range(5), timeout=2))
+        except asyncio.TimeoutError:
+            pass
+        else:
+            assert False
+
+
+def test_map_error_without_wait_no_error():
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        bpe.map(lambda _: sleep_forever(), range(5), timeout=2)
+
+
+def test_exception_in_map():
+    def raise_value_error():
+        raise ValueError('dead')
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        try:
+            gen = bpe.map(lambda _: raise_value_error(), range(5))
+            next(gen)
+        except ValueError as exc:
+            assert 'ValueError: dead' in exc.args[0]
+        else:
+            assert False
+
+
+def test_exception_in_result():
+    def raise_value_error():
+        raise ValueError('dead')
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        try:
+            future = bpe.submit(raise_value_error)
+            future.result()
+        except ValueError as exc:
+            assert 'ValueError: dead' in exc.args[0]
+        else:
+            assert False
+
+
+def test_exception_in_exception():
+    def raise_value_error():
+        raise ValueError('dead')
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        try:
+            future = bpe.submit(raise_value_error)
+            future.exception()
+        except ValueError as exc:
+            assert 'ValueError: dead' in exc.args[0]
+        else:
+            assert False
+
+
+def test_no_exception_when_shutdown():
+    def raise_value_error():
+        raise ValueError('dead')
+    with BatchPoolExecutor(project='hail-vdc') as bpe:
+        future = bpe.submit(raise_value_error)
+    try:
```

Comment:
I don't see how this test is different than the one above and where the shutdown comes into play.

Reply:
I've renamed to `when_exiting_context`. Leaving the context just invokes the shutdown code.

I should not raise an error when I leave the context of the BatchPoolExecutor, even if `wait_on_exit=True`. The BatchPoolExecutor will wait for all tasks to complete (possibly with an exception) and then shut everything down. I won't receive that exception until I actually look at the future itself, as I do below. In a previous version of BPE, I had a bug where shutdown actually raised exceptions for failing tasks.